src/System.Slices/System/Buffers/OwnedMemory.cs

```diff
@@ -198,21 +207,41 @@ internal interface IMemory<T> : IKnown
         unsafe bool TryGetPointer(long id, out void* pointer);
     }
 
-    public struct DisposableReservation : IDisposable
+    public struct DisposableReservation<T> : IDisposable
     {
-        IKnown _owner;
+        OwnedMemory<T> _owner;
         long _id;
 
-        internal DisposableReservation(IKnown owner, long id)
+        internal DisposableReservation(OwnedMemory<T> owner, long id)
         {
             _id = id;
             _owner = owner;
-            _owner.AddReference(_id);
+            switch(OwnedMemorySettings.s_mode) {
+                case Mode.RefCount:
+                    ((IKnown)_owner).AddReference(_id);
+                    break;
+                case Mode.LocalRefCount:
+                    ReferenceCounter.AddReference(_owner);
+                    break;
+                case Mode.None:
+                    break;
+            }
         }
 
+        public Span<T> Span => _owner.Span;
```

Comment:
I can do this, but for clarity which paths are considered Dangerous?  For instance, I think Memory<T>.Span is going to be Dangerous, should that also be renamed?
To me only 

```
using(var res = Memory<T>.Reserve()) {
     var s = res.Span;
     ...
}
```

Is the only way to get a Span from a Memory that doesn't have `Dangerous` in the external API?